crates/hyprstream/src/runtime/generation_core.rs

```diff
@@ -48,107 +50,169 @@ impl SamplingParams {
-            let next_token = self.engine.sample_token_with_params(
-                &logits,
-                params,
-                &input_ids,
-            )?;
-
-            // Step 3: Validate token BEFORE adding to context
-
-            // Check if token is beyond the tokenizer's vocabulary
-            // Tokens beyond vocab size indicate a bug in sampling or model configuration
-            let vocab_size = self.engine.get_vocab_size();
-            if vocab_size > 0 && next_token >= vocab_size {
-                tracing::error!(
-                    "Generated token {} exceeds vocabulary size {}. This indicates a bug in model initialization or sampling.",
-                    next_token, vocab_size
-                );
-                return Err(anyhow::anyhow!(
-                    "Generated out-of-vocabulary token {}: exceeds vocab size {}",
-                    next_token,
-                    vocab_size
-                ));
-            }
+        // Clear and prepare our generated tokens buffer
+        self.generated_tokens.clear();
+        self.generated_tokens.reserve(request.max_tokens);
 
-            // Check for EOS token from model config
-            if self.engine.is_eos_token(next_token) {
-                tracing::debug!("EOS token detected: {}", next_token);
-                break;
-            }
+        // Create DecodeStream for incremental decoding (O(1) per token!)
+        let tokenizer = self.engine.get_tokenizer()?;
+        let mut decode_stream = tokenizer.decode_stream(false); // Don't skip special tokens
 
-            // Step 4: Add token to sequence ONLY after validation
-            input_ids.push(next_token as i64);
-            tokens_generated += 1;
-
-            // Step 5: Decode new text incrementally using custom decoder
-            // Note: Using IncrementalUtf8Decoder due to tokenizers 0.20.4 DecodeStream bugs
-            // TODO: Upgrade to tokenizers 0.22.1+ to use native DecodeStream API for O(1) decoding
-            let new_text = self.decoder.push_token_simple(next_token as i64, |tokens| {
-                let mut full_ids = Vec::with_capacity(prompt_len + tokens.len());
-                full_ids.extend(&input_ids[..prompt_len]);
-                full_ids.extend(tokens);
-                self.engine.detokenize(&full_ids[prompt_len..])
-            })?;
-
-            // Step 6: Call callback with new text (if any)
-            if !new_text.is_empty() {
-                match callback(&new_text)? {
-                    CallbackControl::Continue => {},
-                    CallbackControl::Stop => break,
+        // Clear KV cache before generation to prevent context pollution from previous runs
+        self.engine.clear_kv_cache();
+
+        // Notify callback that generation is starting
+        callback.on_start().await;
+
+        // Create the generation future
+        let generation_future = async {
+            for i in 0..request.max_tokens {
```

Comment:
Should this be switched to a while loop so filtered tokens don't comsume the quota? 

```
let generation_future = async {
let vocab_size = self.engine.get_vocab_size();
let mut i: usize = 0;
while tokens_generated < request.max_tokens {
```